pkg/controller/workloadspread/update_pod_deletion_cost.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		return err
+	}
+	err = r.updateDeletionCostForSubsetPods(ws, subset.Name, negativePods, strconv.Itoa(wsutil.PodDeletionCostNegative*(subsetIndex+1)))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcileWorkloadSpread) updateDeletionCostForSubsetPods(ws *appsv1alpha1.WorkloadSpread,
+	subsetName string, pods []*corev1.Pod, deletionCostStr string) error {
+	for _, pod := range pods {
+		if err := r.updatePodDeletionCost(ws, pod, deletionCostStr); err != nil {
+			r.recorder.Eventf(ws, corev1.EventTypeWarning,
+				"UpdatePodDeletionCostFailed",
+				"WorkloadSpread %s/%s failed to update deletion-cost annotation to %d for Pod %s/%s in subset %s",
+				ws.Namespace, ws.Name, deletionCostStr, pod.Namespace, pod.Name, subsetName)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileWorkloadSpread) updatePodDeletionCost(ws *appsv1alpha1.WorkloadSpread,
+	pod *corev1.Pod, deletionCostStr string) error {
+	clone := pod.DeepCopy()
+	refresh := false
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if refresh {
+			key := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+			if err := r.Get(context.TODO(), key, clone); err != nil {
+				klog.Errorf("error getting Pod (%s/%s) from client", pod.Namespace, pod.Name)
+				return err
+			}
+		}
+
+		updated := updatePodAnnotation(clone, wsutil.PodDeletionCostAnnotation, deletionCostStr)
+		if !updated {
+			return nil
+		}
+
+		updateErr := r.Update(context.TODO(), clone)
+		if updateErr == nil {
+			klog.V(3).Infof("WorkloadSpread (%s/%s) updated deletion-cost annotation to %s for Pod (%s/%s) successfully",
+				ws.Namespace, ws.Name, deletionCostStr, pod.Namespace, pod.Name)
+			return nil
+		}
+		refresh = true
+		return updateErr
+	})
+	return err
+}
+
+func updatePodAnnotation(pod *corev1.Pod, annotationKey, annotationValue string) bool {
+	podAnnotation := pod.GetAnnotations()
+	oldValue, exist := podAnnotation[annotationKey]
+	// annotation has been set
+	if exist && annotationValue == oldValue {
+		return false
+	}
+	// keep the original condition.
+	if !exist && annotationValue == "0" {
+		return false
+	}
+
+	if podAnnotation == nil {
+		podAnnotation = make(map[string]string)
+	}
+	podAnnotation[annotationKey] = annotationValue
+	pod.SetAnnotations(podAnnotation)
+
+	return true
+}
+
+func sortDeleteIndexes(pods []*corev1.Pod) []int {
+	waitDeleteIndexes := make([]int, 0, len(pods))
+	for i := 0; i < len(pods); i++ {
+		waitDeleteIndexes = append(waitDeleteIndexes, i)
+	}
+
+	// Sort Pods with default sequence
+	//	- Unassigned < assigned
+	//	- PodPending < PodUnknown < PodRunning
+	//	- Not ready < ready
+	//	- Been ready for empty time < less time < more time
+	//	- Pods with containers with higher restart counts < lower restart counts
+	//	- Empty creation time pods < newer pods < older pods
+	sort.Slice(waitDeleteIndexes, func(i, j int) bool {
```

Comment:
consider replace slice with slicestable, so that pods deletion cost will be changed less frequently

Reply:
Got it, `slicestable` is a stable sort, thus it is more suitable.